sql/src/main/java/io/crate/planner/operators/Order.java

```diff
@@ -101,9 +102,45 @@ public ExecutionPlan build(PlannerContext plannerContext,
         return executionPlan;
     }
 
+    @Override
+    public LogicalPlan tryOptimize(@Nullable LogicalPlan pushDown) {
+        if (pushDown instanceof Order) {
+            // We can combine the order being pushed with this Order
+            OrderBy mergedOrderBy = ((Order) pushDown).orderBy.merge(orderBy);
+            Order mergedOrder = new Order(source, mergedOrderBy);
+            return source.tryOptimize(mergedOrder);
+        }
+        if (pushDown != null) {
+            // already pushing down something else
+            return null;
+        }
+        // try pushing down this Order (if possible)
+        LogicalPlan optimize = source.tryOptimize(this);
+        if (optimize == null) {
+            return super.tryOptimize(null);
+        }
+        return optimize;
+    }
+
     @Override
     protected LogicalPlan newInstance(LogicalPlan newSource) {
-        return new Order(newSource, orderBy);
+        // Replaces any Symbols which were part of the old orderBy
+        // with the Symbols of the new orderBy at the same position.
+        // This ensures that the orderBy is performed performed
+        // correctly when it's sources have changed.
+        // See tryOptimize of Union
+        List<Symbol> oldOutputs = source.outputs();
+        List<Symbol> newOutputs = newSource.outputs();
+        OrderBy newOrderBy = this.orderBy.copyAndReplace(
+            (symbol) -> {
+                int idx = oldOutputs.indexOf(symbol);
+                if (idx != -1) {
+                    return newOutputs.get(idx);
+                }
+                return symbol;
```

Comment:
If I understand this correctly this case should never happen? Can we raise an error here?

Reply:
Throwing an error now.